src/core/AutomationPattern.cpp

```diff
@@ -178,9 +178,19 @@ const AutomationPattern::objectVector& AutomationPattern::objects() const
 
 MidiTime AutomationPattern::timeMapLength() const
 {
-	if( m_timeMap.isEmpty() ) return 0;
+	MidiTime one_bar = MidiTime(1, 0);
+	if (m_timeMap.isEmpty())
+	{
+		return one_bar;
+	}
 	timeMap::const_iterator it = m_timeMap.end();
-	return MidiTime( MidiTime( (it-1).key() ).nextFullBar(), 0 );
+	tick_t last_tick = static_cast<tick_t>((it-1).key());
+	bar_t last_bar = qMax(0, MidiTime(last_tick).nextFullBar() - 1);
+	if (last_bar == 0 && last_tick == 0)
+	{
+		return one_bar;
+	}
```

Comment:
I think you could write:
```suggestion
	tick_t last_tick = static_cast<tick_t>((it-1).key());
	if (last_tick == 0)
	{
		return one_bar;
	}
	bar_t last_bar = MidiTime(last_tick).nextFullBar() - 1;
```
Ticks must be absolute for `MidiTime(last_tick).nextFullBar()` to return the right bar, so you may as well just check if we're at zero ticks and skip checking `last_bar == 0`.

This lets us short circuit a little faster and makes the qMax call redundant (`nextFullBar` can't be less than zero if `last_tick >= 1`

Reply:
Sometimes one overthinks the solution. Good catch. 👍